### ISSUE TITLE
Two source files with the same content may "collide" in direct cache

### ISSUE THREAD

**Issue by andrey-malets** [abyss7/dist-clang#65]
Source root can have to identical files in the different directories, which may compile into different object code: for example, if we do `#include "file.c"` from two different locations, this can mean two different object files must be generated, because contents of `file.c` from two different locations may not be the same.

When checking direct cache with clangd, we only look at _(source file content, command line, version)_ triple, where _command line_ does not include source file name. And if two source files compile with the same version and flags (which does not look awkward in a typical project), clangd may cache object code from compiling the first file, and then "hit" the cache searching for matching content from the second file, producing wrong object code for it.

I created simple example and [attached](https://github.com/abyss7/dist-clang/files/377069/clang-test.zip) it. Reproduction is as follows:
```
malets@ub:/tmp/clang-test$ make clean
rm -f a/file.o a/file.d b/file.o b/file.d lib.so
malets@ub:/tmp/clang-test$ make
cc -c -MD -MF a/file.d -o a/file.o a/file.c
cc -c -MD -MF b/file.d -o b/file.o b/file.c
ld -shared -o lib.so a/file.o b/file.o
malets@ub:/tmp/clang-test$ make clean
rm -f a/file.o a/file.d b/file.o b/file.d lib.so
malets@ub:/tmp/clang-test$ make CC=/usr/bin/dist-clang/clang
/usr/bin/dist-clang/clang -c -MD -MF a/file.d -o a/file.o a/file.c
/usr/bin/dist-clang/clang -c -MD -MF b/file.d -o b/file.o b/file.c
ld -shared -o lib.so a/file.o b/file.o
b/file.o: In function `f':
a/file.c:(.text+0x0): multiple definition of `f'
a/file.o:a/file.c:(.text+0x0): first defined here
make: *** [lib.so] Error 1
```

**Comment by andrey-malets** [comment]
I presented a draft of possible solution in #66: while doing direct cache search and going through a list of dependencies, we add an additional check that the input file is contained within this list. If the input file name is missing from the list, that that cache entry must have been constructed from a different source file (at least, from a different absolute/relative location as seen by compiler command line), and we avoid hitting it. This leads to the following:

All the entries with the same source file content "contend" for direct cache entry: for example, if we compile the attached code with patched version from clean checkout multiple times, then neither `a/file.c` nor `b/file.c` will hit direct cache, because after compiling `a/file.c` the cache entry will "point" to that file as a dependency, and the cache will miss compiling `b/file.c` thereafter, and vice versa: after compiling `b/file.c` the entry will point to that file, and if we begin another try with `a/file.c`, we again miss direct cache.

However, both items will be served by simple cache just fine, and if we do not "repoint" direct cache entry by compiling another file with the same content in-between, then every entry has a chance to hit direct cache too (e.g. double compilation of `a/file.c` alone will give direct cache hit as before).

I think this rare direct cache misses for source files with the same content are worth paying for possible false positives.

**Comment by andrey-malets** [comment]
Fixed in #67.
